File: src/main/kotlin/com/xs0/gqlktx/types/gql/GInputObjType.java
package com.xs0.gqlktx.types.gql;

import com.xs0.gqlktx.QueryException;
import com.xs0.gqlktx.schema.builder.TypeKind;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;

import java.util.HashSet;
import java.util.Map;

public class GInputObjType extends GFieldedType {
    public GInputObjType(String name, Map<String, GField> fields) {
        super(name, fields);
    }

    @Override
    public TypeKind getKind() {
        return TypeKind.INPUT_OBJECT;
    }

    @Override
    public boolean validAsArgumentType() {
        return true;
    }

    @Override
    public void coerceValue(JsonArray raw, int index, JsonArray out) throws QueryException {
        JsonObject rawObj;
        try {
            rawObj = raw.getJsonObject(index);
        } catch (ClassCastException e) {
            throw new QueryException("Expected a JSON object value for type " + getGqlTypeString());
        }

        if (rawObj == null) {
            out.addNull();
        } else {
            out.add(makeCoercedObj(rawObj));
        }
    }

    @Override
    public void coerceValue(JsonObject raw, String key, JsonObject out) throws QueryException {
        JsonObject rawObj;
        try {
            rawObj = raw.getJsonObject(key);
        } catch (ClassCastException e) {
            throw new QueryException("Expected a JSON object value for type " + getGqlTypeString());
        }

        if (rawObj == null) {
            out.putNull(key);
        } else {
            out.put(key, makeCoercedObj(rawObj));
        }
    }

    private JsonObject makeCoercedObj(JsonObject rawObj) throws QueryException {
        if (!getFields().keySet().containsAll(rawObj.fieldNames())) {
            HashSet<String> names = new HashSet<>(rawObj.fieldNames());
            names.removeAll(fields.keySet());

            throw new QueryException("Unknown field(s) in value of type " + getGqlTypeString() + ": " + names);
        }

        JsonObject coerced = new JsonObject();
        for (Map.Entry<String, GField> entry : fields.entrySet()) {
            entry.getValue().getType().coerceValue(rawObj, entry.getKey(), coerced);
        }
        return coerced;
    }

    @Override
    public void toString(StringBuilder sb) {
        sb.append("input ").append(getName()).append(" {\n");
        dumpFieldsToString(sb);
        sb.append("}\n");
    }
}

File: src/main/kotlin/com/xs0/gqlktx/types/gql/GInputObjType.kt
package com.xs0.gqlktx.types.gql

import com.xs0.gqlktx.QueryException
import com.xs0.gqlktx.dom.*
import com.xs0.gqlktx.schema.builder.TypeKind
import com.xs0.gqlktx.schema.intro.GqlIntroInputValue

class GInputObjType(name: String, val inputFields: Map<String, GInputField>, description: String?) : GBaseType(name, description) {

    override val kind: TypeKind
        get() = TypeKind.INPUT_OBJECT

    override fun coerceValue(raw: Value): ValueObject {
        if (raw !is ValueObject)
            throw QueryException("Expected an object value for type $gqlTypeString")

        if (!inputFields.keys.containsAll(raw.elements.keys)) {
            val unknown = raw.elements.keys - inputFields.keys
            throw QueryException("Unknown field(s) in value of type $gqlTypeString: $unknown")
        }

        val coerced = LinkedHashMap<String, ValueOrNull>()
        for ((key, field) in inputFields) {
            if (key !in raw.elements.keys)
                continue

            when (val rawValue: ValueOrVar = raw.elements.getValue(key)) {
                is ValueNull -> {
                    if (field.type.kind == TypeKind.NON_NULL)
                        throw QueryException("Null value not allowed for field $key")

                    coerced[key] = rawValue
                }

                is Variable -> {
                    throw IllegalStateException("Variable?!?") // this should never happen
                }

                is Value -> {
                    coerced[key] = field.type.coerceValue(rawValue)
                }
            }
        }

        return ValueObject(coerced)
    }

    val inputFieldsForIntrospection: List<GqlIntroInputValue>
        get() {
            val res = ArrayList<GqlIntroInputValue>(inputFields.size)
            for ((_, value) in inputFields)
                res.add(GqlIntroInputValue(value))

            return res
        }

    override fun toString(sb: StringBuilder) {
        sb.append("input ").append(name).append(" {\n")
        for ((_, value) in inputFields) {
            sb.append("  ")
            value.toString(sb)
            sb.append("\n")
        }
        sb.append("}\n")
    }
}
